### PR TITLE
Move nickname functionality to NicknameServlet

### PR DESCRIPTION
**Pull request by n-wach**
Also, nicknames are checked in real time as they are typed into the nickname form.  The submit button shows any errors with the name (taken, too long, etc.).  Nicknames must be accepted before they can be submitted and set.

### CODE REVIEW

**Comment by n-wach** [comment]
Also the nickname checkbox of #42

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.auto.value.AutoValue;
+import java.util.Optional;
+
+/** Contains a response from the {@code NicknameServlet}. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:20]: Can you expand on this comment. Maybe: Contains a response for a nickname proposed to the {@code NicknameServlet}

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+@AutoValue
+public abstract class NicknameResponse {
+  public enum Status {
+    ACCEPTABLE,
+    UNACCEPTABLE,
+  }
+
+  /**
+   * Creates an unaccepting response for a nickname with the given reason.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:29]:
```suggestion
   * Creates a rejected response for {@code nickname} with the given {@code reason}.
```

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+   */
+  public static NicknameResponse reject(String nickname, String reason) {
+    return create(nickname, Status.UNACCEPTABLE, Optional.of(reason));
+  }
+
+  /**
+   * Creates an accepting response for a nickname with no reason.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:36]:
```suggestion
   * Creates a valid response for {@code nickname} with no reason.
```

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  /**
+   * Returns the {@code Status} of the response.
+   *
+   * <p>Whether the nickname is acceptable or not.
+   */
+  public abstract Status status();
+
+  /**
+   * Returns the reason for the {@code status}.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:57]:
```suggestion
   * Returns the reason for the {@code Status}.
```

File: portfolio/src/main/java/com/google/sps/servlets/UserServlet.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  public void init() {
+    userService = UserServiceFactory.getUserService();
+    auth = Authenticator.getAuthenticator();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    final LoginStatus loginStatus;
+    if (auth.isUserLoggedIn()) {
+      String logoutUrl = userService.createLogoutURL("/");
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/UserServlet.java:47]: is this the home path? This should be a private static const so it is clear what it represents.

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -161,26 +158,40 @@ function makeSubmission(topic) {
...
+
+    const request = await fetch(url);
+    const nicknameResponse = await request.json();
+
+    if (nicknameResponse.status == NICKNAME_ACCEPTABLE) {
+      submit.disabled = false;
+      submit.value = 'Submit';
+    } else {
+      submit.disabled = true;
+      submit.value = nicknameResponse.reason.value;
```
Comment (billie-cat) [inline portfolio/src/main/webapp/script.js:192]: Return early and then you can get rid of the else statement.

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 7]
```diff
@@ -33,28 +32,30 @@
 
 
 /**
- * Servlet that manages logging in and out of Nathan's site.
+ * Servlet that can check a user's nickname, and set it once acceptable.
```
[the remark this answers is not in the document]
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:35]: Done

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 8]
```diff
@@ -86,34 +88,46 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
   }
 
-  private boolean isValidNickname(String nickname) {
-    if (nickname.length() < MIN_NICKNAME_LEN || nickname.length() > MAX_NICKNAME_LEN) {
-      return false;
+  /**
+   * Returns a {@code NicknameResponse} based on the provided nickname.
+   *
+   * <p>This checks a number of qualities of the nickname to determine if it is acceptable. For
+   * instance, nicknames must be alphanumeric, lowercase, unique, etc.
```
[the remark this answers is not in the document]
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:108]: Sure, done

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 1 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:20]: Done

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 2 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:29]: Done

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 3 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:36]: Done, used `accepted` instead of `valid` to match method name

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 4 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:57]: This refers to the field `status` not the enum, should it still be capitalized?

File: portfolio/src/main/java/com/google/sps/servlets/UserServlet.java (hunk 5 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/UserServlet.java:47]: done

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (n-wach) [reply portfolio/src/main/webapp/script.js:192]: also inverted conditional (checking for UNACCEPTABLE) now so the success condition is at the end

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ */
+@AutoValue
+public abstract class NicknameResponse {
+  public enum Status {
+    ACCEPTABLE,
+    UNACCEPTABLE,
+  }
+
+  /**
+   * Creates a rejected response for a {@code nickname} with the given {@code reason}.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:35]: nit: it is not "a {@code nickname}" it is "the {@code nickname}"

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+   */
+  public static NicknameResponse reject(String nickname, String reason) {
+    return create(nickname, Status.UNACCEPTABLE, Optional.of(reason));
+  }
+
+  /**
+   * Creates an accepted response for a {@code nickname} with no reason.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:42]: nit: same here. Not "a {@code nickname}"

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 4 above)
Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:63]: So you should never refer to internal state in your JavaDocs. Now that I think about it. You should not be returning a free form text. You should expand on STATUS to have more than 2 status.

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+/**
+ * Contains a response for a nickname proposed to the {@code NicknameServlet}.
+ *
+ * <p>A client may query the {@code NicknameServlet} to ask if a given nickname is allowed.
+ * The servlet will send a {@code NicknameResponse} (as JSON) back to client, which can use it
+ * to display an error or enable submission.
+ */
+@AutoValue
+public abstract class NicknameResponse {
+  public enum Status {
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:29]: Expand on types of status to include:

VALID,
INVALID_NON_ALPHANUMERIC,
INVALID_NON_LOWERCASE,
INVALID_TOO_SHORT,
INVALID_TOO_LONG,
INVALID_RESERVED,
INVALID_TAKEN}

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:29]: This way you can obtain the reason it was rejected without having to call reason().

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 10 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:42]: Changed

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 9 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:35]: Changed

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 11 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:29]: Chatted about this and have implemented a better solution

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+  private static NicknameResponse create(boolean valid, Optional<String> reason) {
+    return new AutoValue_NicknameResponse(valid, reason);
+  }
+
+  /**
+   * Returns {@code true} if the nickname is valid.
+   */
+  public abstract boolean isValid();
+
+  /**
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:52]: You do not need both isValid and isInvalid. I would remove isInvalid.

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+   * Returns {@code true} if the nickname is invalid.
+   */
+  public boolean isInvalid() {
+    return !isValid();
+  }
+
+  /**
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:59]: The summary comment should include any special return values. I would rephrase this as:

/**
 * Returns the reason the nickname is invalid. If the nickname is valid, then returns {@code Optional.empty()}.
 */

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 14]
```diff
@@ -123,12 +144,18 @@ private boolean isValidNickname(String nickname) {
...
 
-    return true;
+    return NicknameResponse.valid();
   }
 
+  /**
+   * Sets the nickname of the current user, and saves it to the datastore.
+   *
+   * <p>This does not check if the nickname is valid, and should only be used on a nickname that
+   * was accepted by {@code queryNickname}.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:157]:
```suggestion
   * was accepted by {@code queryNickname()}.
```

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 12 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java:52]: Done

File: portfolio/src/main/java/com/google/sps/data/NicknameResponse.java (hunk 13 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/data/NicknameResponse.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java (hunk 14 above)
Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java]: Done

File: portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java [hunk 15]
```diff
@@ -86,34 +88,53 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
     }
 
-    // Nickname cannot be "anonymous", as this is reserved for anonymous comments.
-    if (nickname.toLowerCase().equals(NICKNAME_ANONYMOUS)) {
-      return false;
+    if (RESERVED_NAMES.contains(nickname)) {
+      // Some nicknames are reserved.  For instance, nickname cannot be "anonymous", as this is
+      // reserved for anonymous comments.
+      return NicknameResponse.invalid("Reserved");
     }
```
Comment (f-camp) [inline portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:138]: These checks should be in a different method to reduce the length of this method, also makes it easier to test separately.

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:138]: Renamed `queryNickname` to `validateNickname`

Comment (n-wach) [reply portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java:138]: Chatted about this and renamed method
